scipy/linalg/autosol.py

```diff
@@ -0,0 +1,405 @@
+        sigma = rms(Utb[ur:mn])
+        lambdah = discrep(A, b, U, S, Vt, ur, sigma)
+        x, check = rmslambdah(A, b, U, S, Vt, ur, lambdah)
+    return x, ur, sigma, lambdah
+
+
+def autosolve_nonneg(A, b):
+    """ Computes a nonnegative solution of A*x = b.
+
+    Autosolve_nonneg uses autosolve, above, and iteratively zeroes
+    variables that violates the nonnegativity constraint.
+
+    Parameters
+    ----------
+    Exactly as autosolve above
+
+    Returns
+    -------
+    Exactly as autosolve above
+    """
+    A = atleast_2d(_asarray_validated(A, check_finite=True))
+    b = atleast_1d(_asarray_validated(b, check_finite=True))
+    m = A.shape[0]
+    n = A.shape[1]
+    if b.shape[0] != m:
+        raise LinAlgError("Inconsistent array sizes.")
+    z = np.zeros(n)
+    if np.count_nonzero(A) == 0:
+        return z, 0, 0.0, 0.0
+    if np.count_nonzero(b) == 0:
+        return z, 0, 0.0, 0.0
+
+    xt, ur, sigma, lambdah = autosolve(A, b)
+    # see if unconstrained solution is already non-negative
+    if min(xt) >= 0.0:
+        return xt
+    # the approach here is to actually delete columns, for SVD speed,
+    # rather than just zero out columns and thereby complicate the SVD.
+    C = A
+    cols = [0] * n  # list of active column numbers
+    for i in range(1, n):
+        cols[i] = i
+    nn = n
+    for i in range(1, nn):
+        # choose a column to zero
+        p = -1
+        worst = 0.0
+        for j in range(0, nn):
+            if xt[j] < worst:
+                p = j
+                worst = xt[p]
+        if p < 0:
+            break
+        # remove column p and resolve
+        C = np.delete(C, p, 1)
+        cols.pop(p)
+        nn -= 1
+        U, S, Vt = np.linalg.svd(C, full_matrices=False)
+        ms = len(S)
+        ps = np.zeros(ms)
+        for i in range(0, ms):
+            ps[i] = 1.0 / (S[i] + lambdah ** 2 / S[i]) if S[i] > 0.0 else 0.0
+        xt = np.transpose(Vt) @ (np.diag(ps) @ (np.transpose(U) @ b))
+
+    # rebuild full solution vector
+    if xt[0] < 0.0:
+        xt[0] = 0.0  # degenerate case
+    x = np.zeros(n)
+    for j in range(0, nn):
+        x[int(cols[j])] = xt[j]
+    return x, ur, sigma, lambdah
+
+
+def autosolve_rising(A, b):
+    """ Computes a rising (non-descending) solution of A*x = b.
+
+    autosolve_rising uses autosolve, above, and iteratively zeroes
+    columns of A that represent now-redundant variables.
+
+    Parameters
+    ----------
+    Exactly as autosolve above
+
+    Returns
+    -------
+    Exactly as autosolve above
+    """
+    A = atleast_2d(_asarray_validated(A, check_finite=True))
+    b = atleast_1d(_asarray_validated(b, check_finite=True))
+    m = A.shape[0]
+    n = A.shape[1]
+    if b.shape[0] != m:
+        raise LinAlgError("Inconsistent array sizes.")
+    z = np.zeros(n)
+    if np.count_nonzero(A) == 0:
+        return z, 0, 0.0, 0.0
+    if np.count_nonzero(b) == 0:
+        return z, 0, 0.0, 0.0
+
+    C = A  # working copy

```

Comment:
Python doesn't make copies in assignments. You have to make explicit copies via `C = A.copy()`